### PR TITLE
refactor(poetry): Use transforms for dependencies parsing

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Apply more idiomatic approach for Zod schema usage

## Context

- Ref: #23830

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap [hunk 1]
```diff
@@ -377,31 +377,14 @@ exports[`modules/manager/poetry/extract extractPackageFile() extracts multiple d
...
-    {
-      "currentValue": "",
-      "datasource": "pypi",
-      "depName": "dep5",
-      "depType": "dependencies",
-      "managerData": {
-        "nestedVersion": false,
-      },
-      "versioning": "pep440",
-    },
```
Comment (zharinov) [inline lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:404]: This is result from `dep5 = {}` string parsing. I think it's an error to produce value without `skip-reason` for it.

File: lib/modules/manager/poetry/extract.spec.ts [hunk 2]
```diff
@@ -57,7 +57,7 @@ describe('modules/manager/poetry/extract', () => {
     it('extracts multiple dependencies (with dep = {version = "1.2.3"} case)', async () => {
       const res = await extractPackageFile(pyproject2toml, filename);
       expect(res).toMatchSnapshot();
-      expect(res?.deps).toHaveLength(8);
+      expect(res?.deps).toHaveLength(7);
```
Comment (zharinov) [inline lib/modules/manager/poetry/extract.spec.ts:60]: `dep5 = {}` removed from results

File: lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap (hunk 1 above)
Comment (rarkins) [reply lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:404]: If we detect the dependency then can we extract it, even if we skip it? This is generally intuitive for users

Comment (zharinov) [reply lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:404]: Okay

File: lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap [hunk 3]
```diff
@@ -504,7 +493,7 @@ exports[`modules/manager/poetry/extract extractPackageFile() extracts multiple d
       "nestedVersion": false,
     },
     "packageName": "dev-dep2",
-    "skipReason": "unspecified-version",
+    "skipReason": "invalid-version",
```
Comment (rarkins) [inline lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:496]: Isn't unspecified more accurate than invalid?

Comment (zharinov) [reply lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:496]: In this particular case `currentValue` equals to `Invalid version.`, so I think `invalid-version` is more accurate.

Comment (rarkins) [reply lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:496]: Isn't it missing, rather than in valid?

Comment (zharinov) [reply lib/modules/manager/poetry/__snapshots__/extract.spec.ts.snap:496]: image

**Comment by zharinov** [comment]
- Blocked by #23978
- Blocked by #23992

**Review by viceice — COMMENTED** [review]
to many unrelated changes 😉

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/manager/poetry/extract.ts [hunk 4]
```diff
@@ -205,9 +139,10 @@ export async function extractPackageFile(
 
   const extractedConstraints: Record<string, any> = {};
 
-  if (is.nonEmptyString(pyprojectfile?.tool?.poetry?.dependencies?.python)) {
-    extractedConstraints.python =
-      pyprojectfile?.tool?.poetry?.dependencies?.python;
+  const pythonVersion =
+    pyprojectfile?.tool?.poetry?.dependencies?.python?.currentValue;
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/manager/poetry/extract.ts:143]: Yes, because it's only part of the entire schema was refactored here

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/poetry/schema.ts [hunk 5]
```diff
@@ -1,17 +1,152 @@
...
 export const PoetryDependencyRecord = LooseRecord(
   z.string(),
-  z.union([PoetryDependencySchema, z.array(PoetryDependencySchema), z.string()])
-);
+  PoetryValue.transform((dep) => {
+    if (dep.skipReason) {
+      return dep;
+    }
+
+    // istanbul ignore if: normaly should not happen
```
Comment (viceice) [inline lib/modules/manager/poetry/schema.ts:116]: why? is the skip reason normally already be set before?

Comment (zharinov) [reply lib/modules/manager/poetry/schema.ts:116]: Exactly

**Review by rarkins — APPROVED** [review]

**Pull request merged**